src/main/java/com/devcycle/sdk/server/local/LocalBucketing.java

```diff
@@ -0,0 +1,149 @@
+package com.devcycle.sdk.server.local;
+
+import static io.github.kawamuray.wasmtime.WasmValType.I32;
+import static io.github.kawamuray.wasmtime.WasmValType.F64;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.github.kawamuray.wasmtime.*;
+import io.github.kawamuray.wasmtime.Module;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.ByteBuffer;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class LocalBucketing {
+    Store<Void> store; // WASM compilation environment
+    Linker linker; // used to read/write to WASM
+    AtomicReference<Memory> memRef; // reference to start of WASM's memory
+
+    public LocalBucketing() {
+        store = Store.withoutData();
+        linker = new Linker(store.engine());
+        memRef = new AtomicReference<>();
+        InputStream wasmInput = getClass().getClassLoader().getResourceAsStream("bucketing-lib.release.wasm");
+        Module module = null;
+        try {
+            module = Module.fromBinary(store.engine(), wasmInput.readAllBytes()); // compile the file
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        setImportsOnLinker(); // get linker ready to instantiate the file by passing in required imports
+        linker.module(store, "", module); // linker now has a runnable instance of the module
+
+        Memory mem = linker.get(store, "", "memory").get().memory();
+        memRef.set(mem);
+    }
+
+    private Collection<Extern> setImportsOnLinker() {
+        Func dateNowFn = WasmFunctions.wrap(store, F64, () -> {
+            return (double) System.currentTimeMillis();
+        });
+        linker.define("env", "Date.now", Extern.fromFunc(dateNowFn));
+
+        Func consoleLogFn = WasmFunctions.wrap(store, I32, (addr) -> {
+            String message = readWasmString(((Number) addr).intValue());
+            System.out.println("Local Bucketing Log - " + message);
+        });
+        linker.define("env", "console.log", Extern.fromFunc(consoleLogFn));
+
+        Func abortFn = WasmFunctions.wrap(store, I32, I32, I32, I32, (messagePtr, filenamePtr, linenum, colnum) -> {
+            String message = readWasmString(((Number) messagePtr).intValue());
+            String fileName = readWasmString(((Number) filenamePtr).intValue());
+            throw new RuntimeException("Local Bucketing Module Aborted. File " + fileName + ", Line " + linenum + ":" + colnum + " \n Message: " + message + "\n");
+        });
+        linker.define("env", "abort", Extern.fromFunc(abortFn));
+
+        return Arrays.asList(Extern.fromFunc(dateNowFn), Extern.fromFunc(consoleLogFn), Extern.fromFunc(abortFn));
+    }
+
+    private int writeWasmString(String param) {
```

Comment:
i'd call this a `NewWasmString` because we're allocating new memory every time - not just writing to memory.